tests/test_validation.py

```diff
@@ -0,0 +1,141 @@
+
+    input_dict = read_yaml_file(input_file)
+    template_dict = read_yaml_file(template_file)
+
+    # checks everything except predicates
+    input_dict = validate_node(input_dict, template_dict)
+
+    # now that all keywords have some value, we can check predicates
+    check_predicates_node(input_dict, input_dict, template_dict)
+
+    return input_dict
+
+
+def test_validation():
+    input_dict = _helper('overall',
+                         'input.yml',
+                         'template.yml')
+
+    reference = {
+        'title': 'this is an example',
+        'scf': {
+            'functional': 'B3LYP',
+            'max_num_iterations': 20,
+            'some_acceleration': True,
+            'thresholds': {
+                 'some_integral_screening': 0.0002,
+                 'energy': 1e-05
+             },
+            'another_number': 10,
+            'some_complex_number': '0.0+0.0j'
+        }
+    }
+
+    assert input_dict == reference
+
+
+def test_template_errors():
+
+    # keyword without doc
+    with pytest.raises(TemplateError) as e:
+        input_dict = _helper('template_errors',
+                             'input.yml',
+                             'template_no_documentation.yml')
+    assert str(e.value) == "keyword(s) without any documentation: ['a_short_string']"
+
+    # keyword with empty doc
+    with pytest.raises(TemplateError) as e:
+        input_dict = _helper('template_errors',
+                             'input.yml',
+                             'template_empty_documentation.yml')
+    assert str(e.value) == "keyword(s) without any documentation: ['a_short_string']"
+
+    # keyword with invalid predicate
+    with pytest.raises(TemplateError) as e:
+        input_dict = _helper('template_errors',
+                             'input.yml',
+                             'template_invalid_predicate.yml')
+    assert str(e.value) == "error in predicate '0 < len(value) <= undefined' in keyword 'a_short_string'"
+
+
+def test_input_errors():
+
+    # unexpected keyword
+    with pytest.raises(InputError) as e:
+        input_dict = _helper('input_errors',
+                             'input_unexpected_keyword.yml',
+                             'template.yml')
+    assert str(e.value) == "found unexpected keyword(s): {'strange'}"
+
+    # unexpected section
+    with pytest.raises(InputError) as e:
+        input_dict = _helper('input_errors',
+                             'input_unexpected_section.yml',
+                             'template.yml')
+    assert str(e.value) == "found unexpected section(s): {'weird'}"
+
+    # keyword which has not default is not set
+    with pytest.raises(InputError) as e:
+        input_dict = _helper('input_errors',
+                             'input_missing_keyword.yml',
+                             'template.yml')
+    assert str(e.value) == "the following keyword(s) must be set: {'a_short_string'}"
+
+    # type errors
+    for file_name, error in [
```

Comment:
I think the test can be refactored using `parametrize` rather than `for` loops internally. Would you mind adding a note to #25?